### ISSUE TITLE
[Good First Issue] [NNCF] Make NNCF common accuracy aware training code pass mypy checks

### ISSUE THREAD

**Issue by vshampor** [openvinotoolkit/nncf#2492]
This is exactly like https://github.com/openvinotoolkit/nncf/issues/2495 (see the description and tasks there), but the target code path for this one is nncf/common/accuracy_aware_training instead of nncf/common/graph.

**Comment by p-wysocki** [comment]
@vshampor could you please either link the PR to the development section of the issue or grant me some permissions in NNCF repo to do this myself?

![image](https://github.com/openvinotoolkit/nncf/assets/55858107/362ad083-c862-4cb0-9991-52d30b75877b)

I'd prefer the latter since I'm keeping an eye on all GFIs weekly.

**Comment by anzr299** [comment]
@p-wysocki Is this issue being actively worked on? If not, I'm down to work on it!

**Comment by p-wysocki** [comment]
Hi @anzr299, sure! The thing is this issue is in NNCF repository and I do not have correct access rights to assign you. @vshampor could you please do it?

For now let's agree the task is yours, feel free to start working on it and ask questions. We'll figure out the formalities in the background. :)

**Comment by anzr299** [comment]
Sure, thank you very much! I have already started working on it.

**Comment by anzr299** [comment]
Hi @p-wysocki, what is the update regarding the assignment? I was off for some time due to my dissertation submission. I am resuming work on this issue now.

**Comment by p-wysocki** [comment]
Done, please let us know if you have any questions @anzr299. :)

**Comment by anzr299** [comment]
Sure, I am following the current types that are mentioned in some files. For arguments with kwargs and args, I am planning on using "...", is that fine?

**Comment by p-wysocki** [comment]
I found some good tips [here](https://stackoverflow.com/questions/37031928/type-annotations-for-args-and-kwargs), I think you can pick one you'd like to use and it's going to be fine - worst case scenario we'll change it.

**Comment by anzr299** [comment]
Alright, I also had another question regarding some initialization with `None`. The following  Callable functions seem to be initialized with None. When adding their type hinting, should I make it optional between None and their actual types?
`       self._train_epoch_fn = None
        self._validate_fn = None
        self._configure_optimizers_fn = None
        self._dump_checkpoint_fn = None
        self._load_checkpoint_fn = None
        self._early_stopping_fn = None
        self._update_learning_rate_fn = None`

**Comment by anzr299** [comment]
Also, I was wondering if the the `|` method would be preferred over the Optional() method to denote an option between type and None since the latest updates suggest a movement towards the `|` direction but the current NNCF code mostly uses optional().

**Comment by anzr299** [comment]
I am facing a problem in the following section of training_loop.py:
![image](https://github.com/openvinotoolkit/nncf/assets/36276944/84a8dd4b-de1f-4c70-b338-83dbc394850b)
Mypy is giving an error saying the CompressionScheduler doesn't contain current_sparsity_level and current_pruning_level. A similar problem exists in the `_interpolate_compression_step_update` method for `runner.maximal_accuracy_drop`.

**Comment by anzr299** [comment]
@p-wysocki @vshampor I've created a PR #2637
